### ISSUE TITLE
Restarting two workflows caused overlap in outputs files

### ISSUE THREAD

**Issue by ionox0** [DataBiosphere/toil#2589]
I intentionally killed two running workflows (using `bkill` command on our lsf cluster), and attempted to restart them using `--restart`. 

I should note that the cause of this may be that the intended way to end a workflow is the `toil kill` command, and perhaps using `bkill` causes some part of the jobstore or a perhaps the jobStoreFileID to be deleted or cached or corrupted in some way. 

The result of this was that one of the runs has been restarted from files of the other jobstore, which resulted in duplicate outputs at the end of both runs. 

The commands were:
```
toil-cwl-runner \
--logFile /home/johnsoni/Projects/Ian/06302R_06302S/set_4/Project_06302_R/Project-2.0.0/log/cwltoil.log \
--jobStore file:///home/johnsoni/Projects/Ian/06302R_06302S/set_4/Project_06302_R/Project-2.0.0/tmp/jobstore-40fb870c-5612-11e9-addb-ec0d9a88a0ee \
--batchSystem lsf \
--workDir /home/johnsoni/Projects/Ian/06302R_06302S/set_4/Project_06302_R/Project-2.0.0/tmp/ \
--outdir /home/johnsoni/Projects/Ian/06302R_06302S/set_4/Project_06302_R/Project-2.0.0 \
--writeLogs /home/johnsoni/Projects/Ian/06302R_06302S/set_4/Project_06302_R/Project-2.0.0/log \
--logLevel DEBUG \
--stats   \
--cleanWorkDir onSuccess  \
--defaultDisk 10G  \
--preserve-environment PATH PYTHONPATH TOIL_LSF_ARGS  \
--retryCount 2  \
--no-container   \
--defaultMem 10G  \
--disableCaching   \
--maxLogFileSize 20000000000  \
--restart  \
/home/johnsoni/pipeline_2.0.0/ACCESS-Pipeline/workflows/ACCESS_pipeline.cwl \
inputs_group_1.yaml
```

and
```
toil-cwl-runner \
--logFile /home/johnsoni/Projects/Ian/06302R_06302S/set_4/Project_06302_R/Project-group_3-2.0.0/log/cwltoil.log \
--jobStore file:///home/johnsoni/Projects/Ian/06302R_06302S/set_4/Project_06302_R/Project-group_3-2.0.0/tmp/jobstore-0e487c9a-561f-11e9-84eb-8cdcd4acd360 \
--batchSystem lsf \
--workDir /home/johnsoni/Projects/Ian/06302R_06302S/set_4/Project_06302_R/Project-group_3-2.0.0/tmp/ \
--outdir /home/johnsoni/Projects/Ian/06302R_06302S/set_4/Project_06302_R/Project-group_3-2.0.0 \
--writeLogs /home/johnsoni/Projects/Ian/06302R_06302S/set_4/Project_06302_R/Project-group_3-2.0.0/log \
--logLevel DEBUG \
--stats   \
--cleanWorkDir onSuccess  \
--defaultDisk 10G  \
--preserve-environment PATH PYTHONPATH TOIL_LSF_ARGS  \
--retryCount 2  \
--no-container   \
--defaultMem 10G  \
--disableCaching   \
--maxLogFileSize 20000000000  \
--restart  \
/home/johnsoni/pipeline_2.0.0/ACCESS-Pipeline/workflows/ACCESS_pipeline.cwl \
inputs_group_3.yaml
```

Thank you!

**Comment by ionox0** [comment]
This sounds similar to what is happening in this issue (same stacktrace):
https://github.com/DataBiosphere/toil/issues/817
```
x23.cbio.private 2019-04-08 11:50:50,718 MainThread WARNING toil.leader: Q/p/jobrOammZ    Traceback (most recent call last):
x23.cbio.private 2019-04-08 11:50:50,718 MainThread WARNING toil.leader: Q/p/jobrOammZ      File "/home/johnsoni/virtualenvs/pipeline_2.0.0/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/worker.py", line 298, in workerScript
x23.cbio.private 2019-04-08 11:50:50,718 MainThread WARNING toil.leader: Q/p/jobrOammZ        job = Job._loadJob(jobGraph.command, jobStore)
x23.cbio.private 2019-04-08 11:50:50,718 MainThread WARNING toil.leader: Q/p/jobrOammZ      File "/home/johnsoni/virtualenvs/pipeline_2.0.0/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/job.py", line 906, in _loadJob
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ        return cls._unpickle(userModule, fileHandle, jobStore.config)
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ      File "/home/johnsoni/virtualenvs/pipeline_2.0.0/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/job.py", line 935, in _unpickle
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ        runnable = unpickler.load()
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ      File "/home/johnsoni/virtualenvs/pipeline_2.0.0/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/job.py", line 1739, in __new__
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ        return cls._resolve(*args)
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ      File "/home/johnsoni/virtualenvs/pipeline_2.0.0/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/job.py", line 1748, in _resolve
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ        with cls._jobstore.readFileStream(jobStoreFileID) as fileHandle:
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ      File "/opt/common/CentOS_6-dev/python/python-2.7.10/lib/python2.7/contextlib.py", line 17, in __enter__
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ        return self.gen.next()
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ      File "/home/johnsoni/virtualenvs/pipeline_2.0.0/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/jobStores/fileJobStore.py", line 349, in readFileStream
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ        self._checkJobStoreFileID(jobStoreFileID)
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ      File "/home/johnsoni/virtualenvs/pipeline_2.0.0/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/jobStores/fileJobStore.py", line 446, in _checkJobStoreFileID
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ        raise NoSuchFileException(jobStoreFileID)
x23.cbio.private 2019-04-08 11:50:50,719 MainThread WARNING toil.leader: Q/p/jobrOammZ    TypeError: __init__() takes at most 3 arguments (4 given)
x23.cbio.private 2019-04-08 11:50:50,720 MainThread WARNING toil.leader: Q/p/jobrOammZ    ERROR:toil.worker:Exiting the worker because of a failed job on host x23.cbio.private
x23.cbio.private 2019-04-08 11:50:50,720 MainThread WARNING toil.leader: Q/p/jobrOammZ    WARNING:toil.jobGraph:Due to failure we are reducing the remaining retry count of job 'CWLWorkflow' Q/p/jobrOammZ with ID Q/p/jobrOammZ to 2
```

I believe what is happening is that this failure is causing a cached jobstore from a different run to be used. If I could get a hint as to the part of the code which reads from such a cached jobstore I would be able to dig deeper.

**Comment by adamnovak** [comment]
The error you posted looks like something is trying to report a `NoSuchFileException`, but is failing to produce the exception argument because it is passing too many arguments. I've opened #2595 to maybe fix that problem, so we can see what file is actually missing.

As for killing wortkflows, did you kill the Toil leader before bkill-ing all the jobs? If you leave the leader running, it is probably going to try and resubmit the jobs, or submit new jobs. If you were using a file jobstore, `toil kill` will kill the Toil leader process and leave the submitted jobs in the queue to finish or not, while just killing the jobs will leave the leader running, potentially submitting new jobs. If you start *another* leader on the same job store without killing the first one, you are going to have a bad time.

As for the original problem with cross-talk between Toil runs, there's definitely no caching of job stores between different Python invocations. Are you sure you are getting cross-talk between different workflows, and not between different runs of the same workflow (i.e. two leaders, because you didn't kill the original leader)?

If you are getting crosstalk between the two different workflows, my instinct would be to blame it on the LSF batch system implementation. It does seem like it properly uses the unique number returned by `bsub` to keep the jobs it owns separate from the jobs it doesn't, but maybe @mr-c can comment on whether it is actually expected to work correctly with multiple simultaneous workflows.

**Comment by ionox0** [comment]
Thanks @adamnovak

I used `bkill` on the leader job, which I now realize is wrong, and `toil kill` should be used. My assumption at this point is that the combination of killing the leader job with worker jobs still running, as well as the fact that I was using the same TMPDIR for both runs, is what caused this issue. I've made changes to keep all temp dirs separate for each run, and will let you know if I encounter something like this again. Thanks very much for the help!
